src/stamps/extends.ts

```diff
@@ -0,0 +1,163 @@
+  await beeDebug.topUpBatch(postageBatchId, amount)
+  const stamp = await beeDebug.getPostageBatch(postageBatchId)
+
+  return stamp
+}
+
+export class ExtendsStampManager extends BaseStampManager {
+  private isBuyingStamp?: boolean = false
+  private topingUpStamps: string[] = []
+  /**
+   * Start the manager in either hardcoded or autobuy mode
+   */
+  async start(config: StampsConfigExtends): Promise<void> {
+    // Extends mode
+    const refreshStamps = async () => this.refreshStampsExtends(config, new BeeDebug(config.beeDebugApiUrl))
+
+    this.stop()
+    await refreshStamps()
+
+    this.interval = setInterval(refreshStamps, config.refreshPeriod)
+  }
+
+  completeTopUp(extendsTypeFeature: 'ttl' | 'capacity', stamp: PostageBatch) {
+    if (extendsTypeFeature === 'ttl') {
+      logger.info('successfully postage stamp TTL extended', { stamp })
+    } else {
+      logger.info('successfully postage stamp capacity extended', { stamp })
+    }
+    // remove stamps from extending stamps array
+    const stampIndex = this.topingUpStamps.findIndex(id => stamp.batchID === id)
+    this.topingUpStamps.splice(stampIndex, 1)
+  }
+
+  async verifyUsableStamps(beeDebug: BeeDebug, ttlMin: number, config: StampsConfigExtends, amount: string) {
+    for (let i = 0; i < this.usableStamps!.length; i++) {
+      const stamp = this.usableStamps![i]
+
+      const minTimeThreshold = ttlMin + config.refreshPeriod / 1000
+
+      if (
+        ttlMin > 0 &&
+        stamp.batchTTL < minTimeThreshold &&
+        !this.topingUpStamps.includes(stamp.batchID) &&
+        amount !== '0'
+      ) {
+        this.topingUpStamps.push(stamp.batchID)
+        logger.info(`extending postage stamp TTL ${stamp.batchID}`)
+
+        try {
+          const stampRes = await topUpStamp(beeDebug, stamp.batchID, amount)
+
+          setTimeout(() => this.completeTopUp('ttl', stampRes), 60000)
+        } catch (e: any) {
+          // error that indicate that 2 stamps are trying to be extended at the same time. Comes out as a warning
+          const errorStampIndex = this.topingUpStamps.indexOf(stamp.batchID)
+          this.topingUpStamps.splice(errorStampIndex, 1)
+          logger.error('failed to topup postage stamp', e)
+        }
+      }
+    }
+  }
+
+  public async refreshStampsExtends(config: StampsConfigExtends, beeDebug: BeeDebug): Promise<void> {
+    let usableStampsExtendsTTL: PostageBatch[] = []
+    let usableStampsExtendsCapacity: PostageBatch[] = []
+    stampCheckCounter.inc()
+    logger.info('checking postage stamps')
+
+    try {
+      const stamps = await beeDebug.getAllPostageBatch()
+
+      const { depth, amount, ttlMin, usageThreshold } = config
+
+      // Get all usable stamps sorted by usage from most used to least
+      usableStampsExtendsTTL = filterUsableStampsExtendsTTL(stamps)
+
+      if (!this.isBuyingStamp) {
+        if (depth > 0 && usableStampsExtendsTTL.length === 0) {
```

Comment:
I don't like this, this is very magicky, because if you don't specify the depth, then it won't buy the initial stamp. I know that @Cafe137 mentioned that he would not expect at all that any stamp would be bought. I am not sure about that myself, IMHO from a UX perspective it makes sense if an initial stamp is bought when not present, but maybe we should make it configurable? Having some flag like `POSTAGE_BUY_INITIAL`? It could even be `true` by default, and if it would be enabled than `depth` would be required? WDYT?

Reply:
I am OK with the initial stamp purchase now that we clarified that autobuy and autoextend are exclusive modes, and autoextend also contains the autobuy feature when all flags are passed.